channels/security/websocket.py

```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
         if scope["type"] != "websocket":
             raise ValueError("You cannot use OriginValidator on a non-WebSocket connection")
         # Extract the Origin header
-        origin_host = None
+        parsed_origin = None
         for header_name, header_value in scope.get("headers", []):
             if header_name == b"origin":
                 try:
-                    origin_host = urlparse(header_value.decode("ascii")).hostname
+                    # Set ResultParse
+                    parsed_origin = urlparse(header_value.decode("ascii"))
                 except UnicodeDecodeError:
                     pass
         # Check to see if the origin header is valid
-        if self.valid_origin(origin_host):
+        if self.valid_origin(parsed_origin):
             # Pass control to the application
             return self.application(scope)
         else:
             # Deny the connection
             return WebsocketDenier(scope)
 
-    def valid_origin(self, origin):
+    def valid_origin(self, parsed_origin):
+        """
+        Checks parsed origin is None.
+
+        Pass control to the validate_origin function.
+
+        Returns ``True`` if validation function was successful, ``False`` otherwise.
+        """
         # None is not allowed
-        if origin is None:
+        if parsed_origin is None:
             return False
-        # Check against our list
-        return validate_host(origin, self.allowed_origins)
+        return self.validate_origin(parsed_origin)
+
+    def validate_origin(self, parsed_origin):
+        """
+        Validate the given origin for this site.
+
+        Check than the origin looks valid and matches the origin pattern in
+        specified list `` allowed_origins``. Any pattern begins with a scheme.
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
+        but it can be omitted. `` * `` matches anything and anything
+        else must match exactly.
+
+        Note. This function assumes that the given origin has a schema, domain and port,
+        but port is optional.
+
+        Returns `` True`` for a valid host, `` False`` otherwise.
+        """
+        return any(
+            pattern == "*" or self.is_allowed_origin(parsed_origin, pattern)
+            for pattern in self.allowed_origins
+        )
+
+    def is_allowed_origin(self, parsed_origin, pattern):
```

Comment:
I would call this `match_allowed_origin` so it's obviously different from `validate_origin`